lib/Analysis/Utility.cpp

```diff
@@ -48,6 +48,12 @@ SmallVector<SmallVector<unsigned>> ReduceOpHelper::getScratchConfigsFast() {
   auto axis = op.axis();
   SmallVector<SmallVector<unsigned>> smemShapes(3);
 
+  auto argLayout = srcTy.getEncoding();
+  auto argLayoutMma = argLayout.dyn_cast<triton::gpu::MmaEncodingAttr>();
+  if (argLayoutMma && argLayoutMma.getVersionMajor() == 2 &&
+      triton::gpu::getWarpsPerCTA(argLayout)[axis] == 1)
+    return {{1, 1}, {1, 1}};
```

Comment:
@ptillet I have some questions about this optimization
1. If I understand correctly, when there is only one warp along the reduction axis, we dont need shared memory to save the intermediate accumulator values. However, later in ReduceOpConversionFast, it still stores to ([code](https://github.com/openai/triton/blob/03419534666dd02ac4eedf19e1e263ec7d745f4d/lib/Conversion/TritonGPUToLLVM/ReduceOpToLLVM.cpp#L403)) and read from ([code](https://github.com/openai/triton/blob/03419534666dd02ac4eedf19e1e263ec7d745f4d/lib/Conversion/TritonGPUToLLVM/ReduceOpToLLVM.cpp#L425)) shared memory. Is it working or am I missing something?
2. If the statement "we dont need shared memory when there is only one warp along the reduction axis" is true, why is it not applied to #blocked encoding?

Reply:
This seems like a bug. The LLVM codegen should be early-exiting in this case :)